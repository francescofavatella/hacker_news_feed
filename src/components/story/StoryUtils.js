export const shortenedText = len => string =>
  string && string.length > len ? string.substring(0, len - 3) + '...' : string;

export const veryShortText = shortenedText(10);
export const shortText = shortenedText(25);
export const longText = shortenedText(500);
