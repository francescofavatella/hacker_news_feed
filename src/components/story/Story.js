import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { longText, shortText } from './StoryUtils';
import styles from './Story.module.css';

const WithLink = props =>
  props.expanded ? (
    <a
      href={props.url}
      className={props.className + ' ' + styles.cursor}
      target={'_blank'}
      onClick={e => e.stopPropagation()}
    >
      {props.children}
    </a>
  ) : (
    <div className={props.className}>{props.children}</div>
  );

export class Story extends Component {
  static propTypes = {
    score: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    storyUrl: PropTypes.string.isRequired,
    upvotingUrl: PropTypes.string.isRequired,
    expanded: PropTypes.bool.isRequired,
  };

  static defaultProps = {
    score: -1,
    title: '',
    author: '',
    text: '',
    upvotingUrl: '',
    storyUrl: '',
    expanded: false,
  };

  createMarkup = str => {
    return { __html: str };
  };
  render() {
    const {
      score,
      title,
      author,
      text,
      upvotingUrl,
      storyUrl,
      expanded,
    } = this.props;

    const storyText = expanded ? text : shortText(text);
    const storyTitle = expanded ? longText(title) : title;

    return (
      <div className={styles.container}>
        <div className={styles.leftColumn}>
          <WithLink expanded={expanded} url={upvotingUrl} className={''}>
            <div className={styles.arrowUp} />
            <div className={styles.score}>{score}</div>
          </WithLink>
        </div>
        <div className={styles.rigthColumn}>
          <WithLink expanded={expanded} url={storyUrl} className={styles.title}>
            {storyTitle}
          </WithLink>
          <div className={styles.author}>{'By ' + author}</div>
          <div
            className={styles.text}
            dangerouslySetInnerHTML={this.createMarkup(storyText)}
          />
        </div>
      </div>
    );
  }
}

export default Story;
