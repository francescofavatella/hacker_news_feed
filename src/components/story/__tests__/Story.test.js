import React from 'react';
import TestRenderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Story from '../Story';
import { shortText, longText, shortenedText } from '../StoryUtils';

const short_str = 'Lorem Ipsum is simply dummy text';
const long_str =
  "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.";

const testSingleProp = ({ component, title, expectedValue, cssClass }) => {
  return describe(title, () => {
    it('renders without crashing', () => {
      shallow(component);
    });

    it('matches snapshot', () => {
      const tree = TestRenderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('creates ' + cssClass + ' field', () => {
      const renderedComponent = shallow(component);
      const componentByCss = renderedComponent.find(cssClass);
      expect(componentByCss).toHaveLength(1);
      expect(componentByCss.props().children).toEqual(expectedValue);
    });
  });
};

const testTextProp = ({ component, title, expectedValue, cssClass }) => {
  return describe(title, () => {
    it('renders without crashing', () => {
      shallow(component);
    });

    it('matches snapshot', () => {
      const tree = TestRenderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('creates ' + cssClass + ' field', () => {
      const renderedComponent = shallow(component);
      const componentByCss = renderedComponent.find(cssClass);
      expect(componentByCss).toHaveLength(1);
      expect(componentByCss.props().dangerouslySetInnerHTML).toEqual({
        __html: expectedValue,
      });
    });
  });
};

describe('Story', () => {
  describe('with no props', () => {
    const component = <Story />;

    it('renders without crashing', () => {
      shallow(component);
    });

    it('matches snapshot', () => {
      const tree = TestRenderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  const score = 100;
  testSingleProp({
    component: <Story score={score} />,
    title: 'with score prop',
    expectedValue: score,
    cssClass: '.score',
  });

  describe('with title prop:', () => {
    const title = 'customTitle';
    testSingleProp({
      component: <Story title={title} />,
      title: 'with a short title prop it should show the same title',
      expectedValue: title,
      cssClass: '.title',
    });

    testSingleProp({
      component: <Story title={long_str} />,
      title:
        'with a very long title and expanded=false props it should show the small title',
      expectedValue: long_str,
      cssClass: '.title',
    });

    testSingleProp({
      component: <Story title={long_str} expanded={true} />,
      title:
        'with a very long title and expanded=true props it should show the long title',
      expectedValue: longText(long_str),
      cssClass: '.title',
    });
  });

  const author = 'customAuthor';
  testSingleProp({
    component: <Story author={author} />,
    title: 'with author prop',
    expectedValue: 'By ' + author,
    cssClass: '.author',
  });

  describe('with text prop:', () => {
    const text = 'customText';
    testTextProp({
      component: <Story text={text} />,
      title: 'with a short text prop it should show the same text',
      expectedValue: text,
      cssClass: '.text',
    });

    testTextProp({
      component: <Story text={long_str} />,
      title:
        'with a very long text and expanded=false props it should show the small text',
      expectedValue: shortText(long_str),
      cssClass: '.text',
    });

    testTextProp({
      component: <Story text={long_str} expanded={true} />,
      title:
        'with a very long text and expanded=true props it should show the full text',
      expectedValue: long_str,
      cssClass: '.text',
    });
  });
});

describe('StoryUtil', () => {
  describe('shortenedText(len)(string)', () => {
    const len = 40;
    const fn = shortenedText(len);
    it('it preserves the short strings', () => {
      expect(fn(short_str).length).toEqual(short_str.length);
    });

    it('it shortens the long strings', () => {
      expect(fn(long_str).length).toEqual(len);
    });
  });
});
