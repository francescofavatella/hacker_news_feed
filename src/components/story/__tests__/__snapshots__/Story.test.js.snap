// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story with author prop matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="leftColumn"
  >
    <div
      className=""
    >
      <div
        className="arrowUp"
      />
      <div
        className="score"
      >
        -1
      </div>
    </div>
  </div>
  <div
    className="rigthColumn"
  >
    <div
      className="title"
    >
      
    </div>
    <div
      className="author"
    >
      By customAuthor
    </div>
    <div
      className="text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
  </div>
</div>
`;

exports[`Story with no props matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="leftColumn"
  >
    <div
      className=""
    >
      <div
        className="arrowUp"
      />
      <div
        className="score"
      >
        -1
      </div>
    </div>
  </div>
  <div
    className="rigthColumn"
  >
    <div
      className="title"
    >
      
    </div>
    <div
      className="author"
    >
      By 
    </div>
    <div
      className="text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
  </div>
</div>
`;

exports[`Story with score prop matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="leftColumn"
  >
    <div
      className=""
    >
      <div
        className="arrowUp"
      />
      <div
        className="score"
      >
        100
      </div>
    </div>
  </div>
  <div
    className="rigthColumn"
  >
    <div
      className="title"
    >
      
    </div>
    <div
      className="author"
    >
      By 
    </div>
    <div
      className="text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
  </div>
</div>
`;

exports[`Story with text prop: with a short text prop it should show the same text matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="leftColumn"
  >
    <div
      className=""
    >
      <div
        className="arrowUp"
      />
      <div
        className="score"
      >
        -1
      </div>
    </div>
  </div>
  <div
    className="rigthColumn"
  >
    <div
      className="title"
    >
      
    </div>
    <div
      className="author"
    >
      By 
    </div>
    <div
      className="text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "customText",
        }
      }
    />
  </div>
</div>
`;

exports[`Story with text prop: with a very long text and expanded=false props it should show the small text matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="leftColumn"
  >
    <div
      className=""
    >
      <div
        className="arrowUp"
      />
      <div
        className="score"
      >
        -1
      </div>
    </div>
  </div>
  <div
    className="rigthColumn"
  >
    <div
      className="title"
    >
      
    </div>
    <div
      className="author"
    >
      By 
    </div>
    <div
      className="text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Lorem Ipsum is simply ...",
        }
      }
    />
  </div>
</div>
`;

exports[`Story with text prop: with a very long text and expanded=true props it should show the full text matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="leftColumn"
  >
    <a
      className=" cursor"
      href=""
      onClick={[Function]}
      target="_blank"
    >
      <div
        className="arrowUp"
      />
      <div
        className="score"
      >
        -1
      </div>
    </a>
  </div>
  <div
    className="rigthColumn"
  >
    <a
      className="title cursor"
      href=""
      onClick={[Function]}
      target="_blank"
    >
      
    </a>
    <div
      className="author"
    >
      By 
    </div>
    <div
      className="text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
        }
      }
    />
  </div>
</div>
`;

exports[`Story with title prop: with a short title prop it should show the same title matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="leftColumn"
  >
    <div
      className=""
    >
      <div
        className="arrowUp"
      />
      <div
        className="score"
      >
        -1
      </div>
    </div>
  </div>
  <div
    className="rigthColumn"
  >
    <div
      className="title"
    >
      customTitle
    </div>
    <div
      className="author"
    >
      By 
    </div>
    <div
      className="text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
  </div>
</div>
`;

exports[`Story with title prop: with a very long title and expanded=false props it should show the small title matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="leftColumn"
  >
    <div
      className=""
    >
      <div
        className="arrowUp"
      />
      <div
        className="score"
      >
        -1
      </div>
    </div>
  </div>
  <div
    className="rigthColumn"
  >
    <div
      className="title"
    >
      Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
    </div>
    <div
      className="author"
    >
      By 
    </div>
    <div
      className="text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
  </div>
</div>
`;

exports[`Story with title prop: with a very long title and expanded=true props it should show the long title matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="leftColumn"
  >
    <a
      className=" cursor"
      href=""
      onClick={[Function]}
      target="_blank"
    >
      <div
        className="arrowUp"
      />
      <div
        className="score"
      >
        -1
      </div>
    </a>
  </div>
  <div
    className="rigthColumn"
  >
    <a
      className="title cursor"
      href=""
      onClick={[Function]}
      target="_blank"
    >
      Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with deskto...
    </a>
    <div
      className="author"
    >
      By 
    </div>
    <div
      className="text"
      dangerouslySetInnerHTML={
        Object {
          "__html": "",
        }
      }
    />
  </div>
</div>
`;
